js/web/lib/wasm/wasm-training-core-impl.ts

```diff
@@ -146,6 +149,163 @@ export const createTrainingSessionHandle =
       }
     };
 
+/**
+ * Prepares input and output tensors by creating the tensors in the WASM side then moving them to the heap
+ * @param trainingSessionId
+ * @param indices for each tensor, the index of the input or output name that the tensor corresponds with
+ * @param tensors list of TensorMetaData
+ * @param tensorHandles should pass in an empty list of numbers; modified in-place by this method & stores the resulting
+ *                      handles of the allocated tensors on the heap
+ * @param inputOutputAllocs modified in-place by this method
+ * @param indexAdd constant to add to the index that is passed to prepareInputOutputTensor
+ */
+const createAndAllocateTensors =
+    (trainingSessionId: number, indices: number[], tensors: Array<TensorMetadata|null>, tensorHandles: number[],
+     inputOutputAllocs: number[], indexAdd: number) => {
+      const wasm = getInstance();
+
+      const count = indices.length;
+      const valuesOffset = wasm.stackAlloc(count * 4);
+
+      // creates the tensors
+      for (let i = 0; i < count; i++) {
+        prepareInputOutputTensor(
+            tensors[i], tensorHandles, inputOutputAllocs, trainingSessionId, indexAdd + indices[i]);
+      }
+
+      // moves to heap
+      let valuesIndex = valuesOffset / 4;
+      for (let i = 0; i < count; i++) {
+        wasm.HEAPU32[valuesIndex++] = tensorHandles[i];
+      }
+
+      return valuesOffset;
+    };
+
+/**
+ * Move output tensors from the heap to an array
+ * @param outputValuesOffset
+ * @param outputCount
+ * @returns
+ */
+const moveOutputToTensorMetadataArr = (outputValuesOffset: number, outputCount: number) => {
+  const wasm = getInstance();
+  const output: TensorMetadata[] = [];
+
+  for (let i = 0; i < outputCount; i++) {
+    const tensor = wasm.HEAPU32[outputValuesOffset / 4 + i];
+
```

Comment:
before you start copying you need to check whether the output tensor was preallocated. If it was then no need to continue:

if (tensor === outputTensorHandles[i]) {
        // output tensor is pre-allocated. no need to copy data.
        output.push(outputTensors[i]!);
        continue;
      }